### ISSUE TITLE
When iget fails because the open operation returns an error, objects go into stale state.

### ISSUE THREAD

**Issue by JustinKyleJames** [irods/irods#5760]
- [x] master
- [x] 4-2-stable
--------

## Bug Report

When iget fails because the open operation returns an error, the object goes into stale state.

### iRODS Version, OS and Version

iRODS 4.2.10.  Tested on Ubuntu 16.

### What did you try to do?

Running iget when the open operation returns an error..

In my particular case, I am detecting that the object is in Glacier so the open plugin operation returns an error indicating that it is in Glacier and is being restored.

### Expected behavior

The error will be returned to the client but the object should stay in good state.

### Observed behavior (including steps to reproduce, if applicable)

The object goes to stale state when the open operation returns an error.  From what I have been told, objects should not go into stale state on read only operations.

```
$ ils -l test1
  rods              0 fujifilmresc            6 2021-08-03.14:34 & test1

$ iget test1 - 
remote addresses: 172.17.0.2 ERROR: getUtil: get error for - status = -721000 S3_FILE_IN_GLACIER_ERR
Level 0: [-]    /github/irods_resource_plugin_s3/s3/s3_operations.cpp:304:irods::error irods_s3::restore_s3_object(irods::plugin_context &) :  status [S3_FILE_IN_GLACIER_ERR]  errno [] -- message [Object is in glacier and is currently bei
ng restored.  Try again later.]

$ ils -l test1                                                                                                                                                     
  rods              0 fujifilmresc            6 2021-08-03.14:34 X test1
```

**Comment by alanking** [comment]
I think we have 2 separate issues here:

1. What should we do with a replica that is in the S3 glacier scenario?

I think that this particular behavior does raise a new case for the status of a replica. This was suggested initially to be a replica status stored in the catalog in the `data_is_dirty` column (a la logical locking). After some more discussion, we think it should be sufficient to detect this error in the server as a special case and handle it appropriately.

What the S3 plugin is doing now - returning `S3_FILE_IN_GLACIER_ERR` and adding an appropriate message to the `rError` stack - is what we want to keep doing. The difference for the new approach would be that the name of the error being returned should be made more generic so that it can be reused for equivalent situations in other storage backends (e.g. `REPLICA_IS_BEING_STAGED` or something like that). The server can then detect this error and choose not to mark the replica as stale.

My follow-up question would be what should we do when a user attempts to overwrite the replica? I imagine that when the system goes to open the replica it fails to do so and the replica is then left in whatever state in which it was found.

2. Should a replica opened in read-only mode have its status updated in general?

I feel like marking the replica as stale in read-only mode is probably not appropriate for many cases (e.g. server connection issues). However, if the repilca is in fact found to be stale when getting/reading the replica, it may be appropriate to mark it as such. For example:
```bash
$ hostname
icat.example.org
$ ilsresc -l demoResc | grep location
location: icat.example.org
$ rm /var/lib/irods/Vault/home/rods/foo
irods@icat:~$ ils -L foo
  rods              0 demoResc          283 2021-08-05.16:57 & foo
        generic    /var/lib/irods/Vault/home/rods/foo
$ ls -l /var/lib/irods/Vault/home/rods/foo
ls: cannot access '/var/lib/irods/Vault/home/rods/foo': No such file or directory
```
In this case, we know that the replica is in fact stale (the file is missing from the vault!), but the catalog doesn't know this yet because nobody has tried to interact with it yet.

What is the expected behavior for `iget foo`? It could leave the replica as it is, or it could update the status to stale because it is known to be not-good.

The only way the status of the replica can be updated is if a client attempts to interact with the data (whether for read or for write). This raises questions of whether the data should be verified in other ways (e.g. checksums, which also requires being able to read the data). So I think this will require more discussion.

---

*TL;DR*: For now, let's just create a more generic error for this scenario, return it from the S3 plugin, and update the server to respond to it appropriately. The staling of replicas when opened in read-only mode will be rolled back and the discussion about how to handle this in the future will be ongoing.

**Comment by trel** [comment]
`REPLICA_IS_BEING_STAGED` seems good.

Other possibilities include...
- `REPLICA_IS_BEING_RESTORED`
- `REPLICA_IS_THAWING`

Already have (but I don't think these are quite right):
- `SYS_REPLICA_INACCESSIBLE`
- `REPLICA_NOT_IN_RESC`

**Comment by alanking** [comment]
The `iget` staling issue has been resolved. Should we leave this issue open for now as we assess options for the Glacier use case and/or open-for-read detecting issues with existing replicas, or should that be a separate issue?

**Comment by JustinKyleJames** [comment]
> My follow-up question would be what should we do when a user attempts to overwrite the replica? I imagine that when the system goes to open the replica it fails to do so and the replica is then left in whatever state in which it was found.

I don't believe that would be a problem.  You can overwrite objects in S3 at any time.  I have tested the scenario where I did an iput on top of an object that was in the middle of restoration and it seemed to work.  I checked back a while later and the new file remained.

Now if someone opens the file in read mode (or read/write) that will fail.  In that case the plugin returns the same error code.

**Comment by trel** [comment]
awaiting new error code in both master and 4-2-stable... and then... the code to sniff that error code.

**Comment by trel** [comment]
@alanking I think we're still needing... "the code to sniff that error code".

**Comment by alanking** [comment]
> The staling of replicas when opened in read-only mode will be rolled back and the discussion about how to handle this in the future will be ongoing.

I think that because we have decided that read-only will not stale replicas (for now), we don't really need to do anything special with this error code. That is, the status of a replica will only change when it is opened for create or write.

If we want to further discuss the possibility of detecting that a replica is in fact stale and updating the catalog to reflect this when opened in read-only mode, I think that will be a bigger change. Perhaps that can be part of this issue, or we start a separate discussion. Thoughts?

**Comment by trel** [comment]
That larger/different discussion should be in a separate issue, agreed - if we need to have it at all.

For this one, since there's no direct action remaining, I'm going to checkbox and close.
